topdown/graphql.go

```diff
@@ -0,0 +1,239 @@
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+	gqltop "github.com/vektah/gqlparser/v2"
+	gqlast "github.com/vektah/gqlparser/v2/ast"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	gqlparser "github.com/vektah/gqlparser/v2/parser"
+	gqlvalidator "github.com/vektah/gqlparser/v2/validator"
+)
+
+// Generate a useful error message from a GraphQL error list.
+// Modeled after the approaches used for similar error list types
+// in docker/containerd.
+func errGraphQLErrorList(builtinName string, errs gqlerror.List) error {
+	switch len(errs) {
+	case 0:
+		return fmt.Errorf("graphql internal error:%s: Error list invoked with zero arguments", builtinName)
+	case 1:
+		return errs[0].Unwrap()
+	default:
+		msg := "errors:\n"
+		for _, err := range errs {
+			msg += builtinName + " error:" + err.Error() + "\n"
+		}
+		return fmt.Errorf(msg)
+	}
+}
+
+// Recursively traverses an AST that has been run through interfaceToValue,
+// and prunes away the fields with null or empty values, and all `Position`
+// structs.
+func pruneIrrelevantGraphQLASTNodes(value ast.Value) ast.Value {
+	// We iterate over the Value we've been provided, and recurse down
+	// in the case of complex types, such as Arrays/Objects.
+	// We are guaranteed to only have to deal with standard JSON types,
+	// so this is much less ugly than what we'd need for supporting every
+	// extant ast type!
+	switch x := value.(type) {
+	case *ast.Array:
+		result := ast.NewArray()
+		// Iterate over the array's elements, and do the following:
+		// - Drop any Nulls
+		// - Drop any any empty object/array value (after running the pruner)
+		for i := 0; i < x.Len(); i++ {
+			vTerm := x.Elem(i)
+			switch v := vTerm.Value.(type) {
+			case ast.Null:
+				continue
+			case *ast.Array:
+				// Safe, because we knew the type before going to prune it.
+				va := pruneIrrelevantGraphQLASTNodes(v).(*ast.Array)
+				if va.Len() > 0 {
+					result = result.Append(ast.NewTerm(va))
+				}
+			case ast.Object:
+				// Safe, because we knew the type before going to prune it.
+				vo := pruneIrrelevantGraphQLASTNodes(v).(ast.Object)
+				if len(vo.Keys()) > 0 {
+					result = result.Append(ast.NewTerm(vo))
+				}
+			default:
+				result = result.Append(vTerm)
+			}
+		}
+		return result
+	case ast.Object:
+		result := ast.NewObject()
+		// Iterate over our object's keys, and do the following:
+		// - Drop "Position".
+		// - Drop any key with a Null value.
+		// - Drop any key with an empty object/array value (after running the pruner)
+		keys := x.Keys()
+		for _, k := range keys {
+			// We drop the "Position" objects because we don't need the
+			// source-backref/location info they provide for policy rules.
+			// Note that keys are quoted strings, like `"example"`.
+			if k.Value.String() == "\"Position\"" {
```

Comment:
Ah, this is exactly what I'm looking for. I remember getting tripped up yesterday by the quotes, so this is perfect.

Reply:
There are a bunch of transformation helpers in the ast package, too. It's a big package 😅 Maybe there's another helper you could use here... 🤔

Reply:
Implemented this change with the second string comparison style you suggested.

Reply:
The transformer system seems interesting, but is going to require me to completely rework how the recursive descent pruning works. I'll give it a go and see if I can't twist my brain around it this afternoon. :sweat_smile:

Reply:
After a few attempts at replacing the recursive descent pruning function with a `Transformer`-based solution, I'm throwing in the towel. `ast/transformer` provides powerful tools for in-place AST modification, but it's definitely painful for my use case of filtering/pruning nodes out selectively.

For now, I think I'm going to leave this alone, since it is working correctly as-is.